extensions/amp-base-carousel/1.0/base-carousel.js

```diff
@@ -17,43 +17,61 @@ import * as Preact from '../../../src/preact';
 import {Arrow} from './arrow';
 import {Scroller} from './scroller';
 import {toChildArray, useRef, useState} from '../../../src/preact';
+import {useMountEffect} from '../../../src/preact/utils';
 
 /**
- * @param {!JsonObject} props
+ * @param {!BaseCarouselDef.Props} props
  * @return {PreactDef.Renderable}
  */
 export function BaseCarousel(props) {
   const {
-    'arrowPrev': arrowPrev,
-    'arrowNext': arrowNext,
-    'children': children,
-    'defaultSlide': defaultSlide,
-    'loop': loop,
+    arrowPrev,
+    arrowNext,
+    children,
+    defaultSlide,
+    loop,
+    slide,
+    onSlideChange,
+    onMount_,
+    setAdvance,
     ...rest
   } = props;
   const childrenArray = toChildArray(children);
   const {length} = childrenArray;
   const [curSlide, setCurSlide] = useState(defaultSlide || 0);
-  const ignoreProgrammaticScroll = useRef(true);
-  const setRestingIndex = (i) => {
-    ignoreProgrammaticScroll.current = true;
-    setCurSlide(i);
-  };
-  const scrollRef = useRef(null);
+  const slideState = slide ?? curSlide;
```

Comment:
I'd look into the issue of controlled/uncontrolled together with `useImpertiveHandle`. I recommend we rip out this code for now and add it in a separate PR. This PR could be dedicated to layout/scroller structure, which is already quite a good set of functionality.